index.js

```diff
@@ -1,183 +1,187 @@
-                no_ready_check: options.noReadyCheck
-            });
-
-            client.on('error', function (err) {
-                log(options, {db: database, action: 'error', err: err.message});
-                if (!callbackCalled) {
-                    callbackCalled = true;
-                    callback(err, client);
-                }
-                client.end(FLUSH_CONNECTION);
-            });
-
-            client.on('ready', function () {
-                client.select(database, function(err/*, res*/) {
-                    if (!callbackCalled) {
-                        callbackCalled = true;
-                        callback(err, client);
-                    }
-                });
-            });
-        },
-
-        destroy: function(client) {
-            client.quit();
+function makePool (options, database) {
+  const factory = {
+    create () {
+      return new Promise((resolve, reject) => {
+        let settled = false;
+
+        const client = redis.createClient(options.port, options.host, {
+          no_ready_check: options.noReadyCheck
+        });
+
+        client.on('error', function (err) {
+          log(options, { db: database, action: 'error', err: err.message });
+
+          if (!settled) {
+            settled = true;
             client.end(FLUSH_CONNECTION);
-        },
-
-        validate: function(client) {
-            return client && client.connected;
-        },
-
-        max: options.max,
-        idleTimeoutMillis: options.idleTimeoutMillis,
-        reapIntervalMillis: options.reapIntervalMillis,
-        returnToHead: options.returnToHead,
-        log: options.log
-    });
-}
 
+            if (err) {
+              return reject(err);
+            }
+            return resolve(client);
+          }
+        });
+
+        client.on('ready', function () {
+          client.select(database, err => {
+            if (!settled) {
+              settled = true;
+
+              if (err) {
+                return reject(err);
+              }
+              return resolve(client);
+            }
+          });
+        });
+      })
+    },
+
+    destroy (client) {
+      return new Promise((resolve, reject) => {
+        client.quit(err => {
+          client.end(FLUSH_CONNECTION);
+          if (err) {
+            return reject(err);
+          }
+          return resolve();
+        });
+      })
+    },
 
-function log(options, what) {
-    if (options.slowPool.log) {
-        console.log(JSON.stringify(_.extend({name: options.name}, what)));
+    validate (client) {
+      return new Promise(resolve => {
+        return resolve(client && client.connected)
+      })
     }
+  }
+
+  const config = {
+    max: options.max,
+    idleTimeoutMillis: options.idleTimeoutMillis,
+    reapIntervalMillis: options.reapIntervalMillis,
+    returnToHead: options.returnToHead
+  }
+
+  return createPool(factory, config);
+}
+
+function log (options, what) {
```

Comment:
_aprovechando que el Pisuerga pasa por Valladolid_

I'd like that the logger were injected as part of the options in the constructor instead of using the `console.log`

I'm gonna work to have a JSON logger (https://www.npmjs.com/package/pino) for every service in CARTO. It'd be super helpful to be able to inject the logger.

Reply:
It is a good thing for a new PR :)

Reply:
no te la cuelo, eh? 😂